### PR TITLE
Simplify exception parsing

### PR DESCRIPTION
**Pull request by jaimergp**
### Description

Investigating a simpler workflow for exception catching, in preparation the new error messages (see #102).

### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [x] Add / update necessary tests?
- [ ] Add / update outdated documentation?

### CODE REVIEW

**Comment by jaimergp** [comment]
One more thing to take into account in the exception parsing:

```
$c onda create -n unused conda=23.5 python --dry-run
conda-forge/osx-arm64                                       Using cache
conda-forge/noarch                                          Using cache
Collect all metadata (repodata.json): done
Solving environment: failed

PackagesNotFoundError: The following packages are not available from current channels:

  - conda
  - 23.5**

Current channels:

  - https://conda.anaconda.org/conda-forge

To search for alternate channels that may provide the conda package you're
looking for, navigate to

    https://anaconda.org

and use the search bar at the top of the page.

```

The spec is space-separated, which is then interpreted as _two_ packages unavailable, but the 2nd one is just the version requested. With a 3-field spec we get three "packages":

```
$ conda create -n unused "conda=23.5=*jaime*" python --dry-run
conda-forge/osx-arm64                                       Using cache
conda-forge/noarch                                          Using cache
Collect all metadata (repodata.json): done
Solving environment: failed

PackagesNotFoundError: The following packages are not available from current channels:

  - conda
  - 23.5
  - *jaime*

Current channels:

  - https://conda.anaconda.org/conda-forge

To search for alternate channels that may provide the conda package you're
looking for, navigate to

    https://anaconda.org

and use the search bar at the top of the page.
```

To be fixed in this PR too.

**Comment by costrouc** [comment]
@jaimergp since this PR depends on libmamba > 1.2.0. defaults now ships 1.3.1 https://anaconda.org/anaconda/libmambapy. Does this need to depend on #133?

**Comment by jaimergp** [comment]
Yea 133 is blocking everything. I am going to rebase that branch while we iterate on those last errors to make some progress here.

**Comment by jaimergp** [comment]
The error on Linux Python 3.10 is a cache loading issue on libmamba side:

```
error    libmamba Could not open lockfile '/opt/conda/pkgs/cache/cache.lock'
warning  libmamba Could not parse state file: Could not load cache state: [json.exception.type_error.302] type must be string, but is null
warning  libmamba Could not remove state file "/opt/conda/pkgs/cache/497deca9.state.json": Permission denied
warning  libmamba Cache file "/home/test_user/.conda/pkgs/cache/497deca9.json" was modified by another program
error    libmamba Could not open lockfile '/opt/conda/pkgs/cache/cache.lock'
warning  libmamba Could not parse state file: Could not load cache state: [json.exception.type_error.302] type must be string, but is null
warning  libmamba Could not remove state file "/opt/conda/pkgs/cache/09cdf8bf.state.json": Permission denied
warning  libmamba Cache file "/home/test_user/.conda/pkgs/cache/09cdf8bf.json" was modified by another program
```

Python-side traceback:

```
Traceback (most recent call last):
  File "/opt/conda-src/conda/exceptions.py", line 1132, in __call__
    return func(*args, **kwargs)
  File "/opt/conda-src/conda/cli/main.py", line 69, in main_subshell
    exit_code = do_call(args, p)
  File "/opt/conda-src/conda/cli/conda_argparse.py", line 92, in do_call
    return getattr(module, func_name)(args, parser)
  File "/opt/conda-src/conda/notices/core.py", line 121, in wrapper
    return func(*args, **kwargs)
  File "/opt/conda-src/conda/cli/main_create.py", line 41, in execute
    install(args, parser, 'create')
  File "/opt/conda-src/conda/cli/install.py", line 264, in install
    unlink_link_transaction = solver.solve_for_transaction(
  File "/opt/conda-src/conda/core/solve.py", line 134, in solve_for_transaction
    unlink_precs, link_precs = self.solve_for_diff(update_modifier, deps_modifier,
  File "/opt/conda-src/conda/core/solve.py", line 177, in solve_for_diff
    final_precs = self.solve_final_state(update_modifier, deps_modifier, prune, ignore_pinned,
  File "/home/test_user/.local/lib/python3.10/site-packages/conda_libmamba_solver/solver.py", line 167, in solve_final_state
    index = LibMambaIndexHelper(
  File "/home/test_user/.local/lib/python3.10/site-packages/conda_libmamba_solver/index.py", line 71, in __init__
    self._index = self._load_channels()
  File "/home/test_user/.local/lib/python3.10/site-packages/conda_libmamba_solver/index.py", line 262, in _load_channels
    set_channel_priorities(self._pool, full_index, self._repos)
  File "/home/test_user/.local/lib/python3.10/site-packages/conda_libmamba_solver/mamba_utils.py", line 162, in set_channel_priorities
    repo = subdir_or_repo.create_repo(pool)
libmambapy.bindings.MambaNativeException: Cache not loaded
```

This is happening in a tight-retry loop to test for determinism, so it looks like it's running into a race condition or something similar. We are also seeing this in https://github.com/conda/conda-libmamba-solver/pull/151. I'd say it's unrelated to this PR and will go away with #65.

So I'll ignore that one for now.

**Comment by jaimergp** [comment]
This is all green now. This is how it looks like:

```
$ conda create -n unused --dry-run "python=*=*cpython*" pypy --solver libmamba
Retrieving notices: ...working... done
conda-forge/noarch                                  11.6MB @ 623.3kB/s 19.2s
conda-forge/linux-aarch64                            8.0MB @ 381.7kB/s 21.5s
Collect all metadata (repodata.json): done
Solving environment: failed

LibMambaUnsatisfiableError: Could not solve for environment specs
================== Experimental satisfiability error messages ==================
You are seeing this because you set `experimental_sat_error_message: true`
Use the following issue to share feedback on this experimental feature
   https://github.com/mamba-org/mamba/issues/2078

====================================== Legacy messages (old) =======================================
Encountered problems while solving:
  - nothing provides python 3.6.9 1_73_pypy needed by pypy-7.3.1-h9f0ad1d_1

The environment can't be solved, aborting the operation
=================================== Experimental messages (new) ====================================
The following packages are incompatible
├─ pypy * * is installable with the potential options
│  ├─ pypy [7.3.0|7.3.1] would require
│  │  └─ python 3.6.9 0_73_pypy, which can be installed;
│  ├─ pypy 7.3.1 would require
│  │  └─ python 3.6.9 1_73_pypy, which does not exist (perhaps a missing channel);
│  ├─ pypy 7.3.1 would require
│  │  └─ python 3.6.9 2_73_pypy, which can be installed;
│  ├─ pypy 7.3.2 would require
│  │  └─ python 3.6.9 3_73_pypy, which can be installed;
│  ├─ pypy 7.3.3 would require
│  │  └─ python 3.6.12 5_73_pypy, which can be installed;
│  ├─ pypy 7.3.3 would require
│  │  └─ python 3.7.9 5_73_pypy, which can be installed;
│  ├─ pypy 7.3.3 would require
│  │  └─ python 3.6.12 4_73_pypy, which can be installed;
│  ├─ pypy [7.3.4|7.3.5] would require
│  │  └─ python 3.7.10 0_73_pypy, which can be installed;
│  ├─ pypy 7.3.5 would require
│  │  └─ python 3.7.10 1_73_pypy, which can be installed;
│  ├─ pypy 7.3.7 would require
│  │  └─ python 3.7.12 0_73_pypy, which can be installed;
│  ├─ pypy 7.3.8 would require
│  │  └─ python 3.8.12 0_73_pypy, which can be installed;
│  ├─ pypy 7.3.8 would require
│  │  └─ python 3.9.10 0_73_pypy, which can be installed;
│  ├─ pypy 7.3.9 would require
│  │  └─ python 3.8.13 0_73_pypy, which can be installed;
│  ├─ pypy 7.3.9 would require
│  │  └─ python 3.9.12 0_73_pypy, which can be installed;
│  ├─ pypy 7.3.11 would require
│  │  └─ python 3.8.16 0_73_pypy, which can be installed;
│  └─ pypy 7.3.11 would require
│     └─ python 3.9.16 0_73_pypy, which can be installed;
└─ python * *cpython* is uninstallable because it conflicts with any installable versions previously reported.
====================================================================================================
```

We should maybe modify the first section because it's a bit misleading. We are setting that variable via the `libmambapy.Context` attribute directly. We might also want to avoid any potential overflow of reports to our mamba friends by pointing to a different issue we control, where we mention the original mamba issue too, for reference.

WDYT @jezdez?

**Comment by jaimergp** [comment]
Last commits implement a message like the following, with a transparent fallback to the legacy if _anything_ (`Exception` caught) fails, just in case.

```
$ conda create -n unused --dry-run "python=*=*cpython*" pypy --solver libmamba
conda-forge/linux-aarch64                                   Using cache
conda-forge/noarch                                          Using cache
Collect all metadata (repodata.json): done
Solving environment: failed

LibMambaUnsatisfiableError: Encountered problems while solving:
  - nothing provides python 3.6.9 1_73_pypy needed by pypy-7.3.1-h9f0ad1d_1

The environment can't be solved, aborting the operation

***************************************************
 conda has set experimental_sat_error_message=true
 You will now see an explanation of the conflicts.
 You can provide feedback to the libmamba team at:
  https://github.com/mamba-org/mamba/issues/2078
***************************************************

The following packages are incompatible
├─ pypy * * is installable with the potential options
│  ├─ pypy [7.3.0|7.3.1] would require
│  │  └─ python 3.6.9 0_73_pypy, which can be installed;
│  ├─ pypy 7.3.1 would require
│  │  └─ python 3.6.9 1_73_pypy, which does not exist (perhaps a missing channel);
│  ├─ pypy 7.3.1 would require
│  │  └─ python 3.6.9 2_73_pypy, which can be installed;
│  ├─ pypy 7.3.2 would require
│  │  └─ python 3.6.9 3_73_pypy, which can be installed;
│  ├─ pypy 7.3.3 would require
│  │  └─ python 3.6.12 5_73_pypy, which can be installed;
│  ├─ pypy 7.3.3 would require
│  │  └─ python 3.7.9 5_73_pypy, which can be installed;
│  ├─ pypy 7.3.3 would require
│  │  └─ python 3.6.12 4_73_pypy, which can be installed;
│  ├─ pypy [7.3.4|7.3.5] would require
│  │  └─ python 3.7.10 0_73_pypy, which can be installed;
│  ├─ pypy 7.3.5 would require
│  │  └─ python 3.7.10 1_73_pypy, which can be installed;
│  ├─ pypy 7.3.7 would require
│  │  └─ python 3.7.12 0_73_pypy, which can be installed;
│  ├─ pypy 7.3.8 would require
│  │  └─ python 3.8.12 0_73_pypy, which can be installed;
│  ├─ pypy 7.3.8 would require
│  │  └─ python 3.9.10 0_73_pypy, which can be installed;
│  ├─ pypy 7.3.9 would require
│  │  └─ python 3.8.13 0_73_pypy, which can be installed;
│  ├─ pypy 7.3.9 would require
│  │  └─ python 3.9.12 0_73_pypy, which can be installed;
│  ├─ pypy 7.3.11 would require
│  │  └─ python 3.8.16 0_73_pypy, which can be installed;
│  └─ pypy 7.3.11 would require
│     └─ python 3.9.16 0_73_pypy, which can be installed;
└─ python * *cpython* is uninstallable because it conflicts with any installable versions previously reported.
```

File: conda_libmamba_solver/solver.py [hunk 1]
```diff
@@ -516,9 +518,32 @@ def _fix_version_field_for_conda_build(spec: MatchSpec):
...
+        First will raise LibmambaUnsatisfiableError.
+        Second will raise PackagesNotFoundError.
+
+        Libmamba can return spec strings in two formats:
+        - With dashes, e.g. package-1.2.3-h5487548_0
+        - à la conda-build, e.g. package 1.2.*
+        - just names, e.g. package
+        """
+
+        def to_matchspec(spec: Union[str, Sequence[str]]):
```
Comment (jezdez) [inline conda_libmamba_solver/solver.py:533]: Something to move out of the method to not dynamically define this method?

File: conda_libmamba_solver/solver.py [hunk 2]
```diff
@@ -527,73 +552,98 @@ def _problems_to_specs_parser(problems: str) -> Mapping[str, MatchSpec]:
...
+                        sections[key].append(line)
+                return "\n".join(
+                    [
+                        *sections["old"],
+                        "",
+                        "***************************************************",
+                        " conda has set experimental_sat_error_message=true",
+                        " You will now see an explanation of the conflicts.",
+                        " You can provide feedback to the libmamba team at:",
+                        "  https://github.com/mamba-org/mamba/issues/2078",
```
Comment (jezdez) [inline conda_libmamba_solver/solver.py:638]:
```suggestion
                        " https://github.com/mamba-org/mamba/issues/2078",
```

Is there prior art for how these block notifications are rendered in conda to make it less block-y?

**Review by jezdez — APPROVED** [review]
Just some minor comments about code structure and UI, no big deal tho

File: conda_libmamba_solver/solver.py (hunk 2 above)
Comment (jaimergp) [reply conda_libmamba_solver/solver.py:638]: I'd say prior art is even block-ier: 

https://github.com/conda/conda-libmamba-solver/blob/dc6b266526a1ab6e07fd01144b6bf420c6f00fd1/conda_libmamba_solver/solver.py#L841-L851

Comment (jaimergp) [reply conda_libmamba_solver/solver.py:639]: The leading space was there to make the URL centered and distinguishable from the paragraph text, but happy to change it.

**Comment by jaimergp** [comment]
Errors in CI are unrelated to this PR (Minio/S3 stuff).

File: conda_libmamba_solver/solver.py (hunk 2 above)
Comment (jezdez) [reply conda_libmamba_solver/solver.py:639]: Nah, it's fine

**Review by jezdez — APPROVED** [review]

**Comment by jaimergp** [comment]
libmamba 1.4.0 has enabled error messages by default so we might need to change the messaging a bit. Looking into it now.

**Comment by jaimergp** [comment]
Yea, the parser broke:

```
DEBUG conda.conda_libmamba_solver.solver:_prepare_problems_message(646): Could not parse 'explain_problems'
Traceback (most recent call last):
  File "/opt/conda-libmamba-solver-src/conda_libmamba_solver/solver.py", line 633, in _prepare_problems_message
    *sections["old"],
KeyError: 'old'
failed
```

**Pull request merged**